### DISCUSSION TITLE
wire:model binding to a boolean property of an Eloquent model doesn't work

### DISCUSSION THREAD

**Discussion by colincameron** [livewire/livewire#7241] [7. Bugs (archived)]
### Livewire version

3.1.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

I have a Livewire component with an Eloquent model as a public property.

```
class ViewAudit extends Component
{

    public Audit $audit;
    ...
```

In the template I have a checkbox with wire:model set to a boolean property of the Audit model.

```
<input type="checkbox" wire:model="audit.hidden">
```

The checkbox doesn't reflect the value of the property correctly, and the JS console in my browser reports the following error on page load, and whenever clicking the checkbox:

```
[Error] TypeError: null is not an object (evaluating 'carry[i]')
	(anonymous function) (livewire.js:347)
	reduce
	dataGet (livewire.js:344)
	get (livewire.js:8888)
	getValue (livewire.js:3292)
	(anonymous function) (livewire.js:3348)
	reactiveEffect (livewire.js:2392)
	effect2 (livewire.js:2367)
	effect (livewire.js:784)
	wrappedEffect (livewire.js:800)
	(anonymous function) (livewire.js:3347)
	flushHandlers (livewire.js:1389)
	stopDeferring (livewire.js:1394)
	deferHandlingDirectives (livewire.js:1397)
	initTree (livewire.js:908)
	(anonymous function) (livewire.js:863)
	forEach
	start (livewire.js:862:95)
	start2 (livewire.js:8112)
	(anonymous function) (livewire.js:9065)
```

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by CyberPunkCodes** [comment]
You are using Livewire version 3, which by default, does not support objects. So `audit.hidden` is not accessible from your blade file. This was supported in Livewire version 2. There was, I guess, some issues with security or potentially exposing sensitive information? I could be wrong on the exact reasoning, but it was disabled in version 3. There is a setting you can set in the config to get that functionality back. I personally think using the setting is chincy and hacky, even though the ability in v2 was nice lol.

Unless your component is SUPER simple and you don't need fancy validations, I recommend always using a Form component. While the form component isn't going to give you access to the object's properties either, it just works better for... well, forms :)

The main benefit I found is the Form component will merge your validation rules AND they can be live. So the `Rule` attributes above your properties and what is defined in the `rules()` method will be merged and allow you to have live validation (ie: using `.live` on the input and it validates as they type). This doesn't work on a regular component. For some reason it only works on the Form component. Almost every time I started with a basic component thinking I didn't need a Form component, I have found myself needing it.

So since v3 doesn't support objects, you have to split out everything you need manually. Something like this:

Main component:

```php
use Livewire\Component;

class EditUser extends Component
{
    public EditUserForm $form;

    public function mount(User $user)
    {
        $this->form->setUser($user);
    }
    
    public function save()
    {
        $this->form->update();
    }
}
```

Form component:

```php
use Livewire\Form;

class EditUserForm extends Form
{
    public User $user;
    
    #[Rule('required|string|min:3|max:60')]
    public $name = '';

    #[Rule('boolean')]
    public $isCool = false;
    
    public function setUser(User $user)
    {
        $this->user = $user;
        $this->name = $user->name;
        $this->isCool = $user->is_cool;
    }
    
    public function update()
    {
        $this->validate();

        $this->user->update(
            $this->only(['name', 'isCool'])
        );
    }
}
```

Then your checkbox needs prefixed with `form`:
```html
<input type="checkbox" wire:model="form.isCool" value="1">
```

Notice how in the main component's mount, I use ` $this->form->setUser($user);` to pass the user object into the form. This isn't necessary if you are creating a new blank record, only when updating an existing one (which it seems like you are doing). Then in the update method on the form, I grab only the fields needed using `only()` so it doesn't try to grab the user object as well.

Hope that helps.
